proposals/TP3-xds-error-propagation.md

```diff
@@ -0,0 +1,122 @@
+TP3: xds-error-propagation
+----
+* Author(s): anicr7
+* Approvers: htuch, markdroth, adisuissa
+* Implemented in: <xDS client, ...>
+* Last updated: 2024-10-25
+
+## Abstract
+
+This proposal introduces enhancements to the [xDS transport protocol](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol), focusing on improved debugging and observability. Specifically, it outlines a mechanism for the control plane to communicate error information directly to xDS clients.
+
+The objective of this proposal is to suggest a way for clients to receive feedback from xDS Management Servers in case of partial/drastic failures without closing any streams or connections.
+
+This proposal includes a new field for each subscribed Resource, called `ResourceError` which will provide detailed information for resource specific issues. The client must use this additional field to obtain notification for resources the xDS Management server couldn’t procure and provide necessary notification to the application. 
+
+## Background
+
+A frequent use case for xDS involves client subscription to multiple resources from the management server. Currently, if the xDS management server cannot provide a subset of these resources, the client experiences a complete loss of visibility. This can occur for various reasons, including but not limited to potential permission issues. This lack of granularity in the response can pose significant operational challenges for general client observability or debugging. 
+
+The xDS protocol does have a way for xDS clients to NACK responses back from the xDS Management server. The NACK response also contains an `error_detail` field which the Management Server can use to extract further information about the rejection. But this NACK’ing mechanism is restricted to the client i.e. there is no way for Management Server to actually convey a notification(Ex: unavailability, permission issues etc) regarding some or all the resource requests that are being subscribed by the client. This eventually leads to the client timeouts which, although it conveys an error back to the application, it misses the actual context for the issue. In most cases the xDS Management Servers might not even be accessible for the client applications to debug these issues without escalation. 
+
+### Related Proposals:
```

Comment:
This should list xRFC TP1, since we're referring to glob collections below.

Reply:
Done